### PR TITLE
MARKET-1564 Totop Component

### PR DESCRIPTION
**Pull request by Fialkaksu**
Totop component allows you to use the up button to scroll up the page or scroll up to a specific page element.

[MARKET-1564](https://bugs.backendless.com/browse/MARKET-1564)

image

### CODE REVIEW

File: components/bl-totop-component/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+<p align="center">
+  <img src="./thumbnail.png" alt="main thumbnail" width="780"/>
+</p>
+
+## Properties
+
+| Property   | Type                                                | Default value  | Logic                      | Data Binding | UI Setting | Description                                                       |
+|------------|-----------------------------------------------------|----------------|----------------------------|--------------|------------|-------------------------------------------------------------------|
+| Position   | *Select* <br> `'bottom-right'` <br> `'bottom-left'` | `bottom-right` | Type Logic                 | NO           | YES        | Controls the position of the up button on the page.               |
+| Element    | *JSON* <br> `{...}`                                 |                | Element To Scroll To Logic | NO           | NO         | The logic to determine the element to which the scroll will be.   |
```
Comment (v-excelsior) [inline components/bl-totop-component/README.md:14]: JSON? What we should pass here? 

I think we could leave Type column for this empty, hide it from settings(as you do it) and add an codeless example screenshot where you pass a component in handler

File: components/bl-totop-component/styles/index.less [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+.bl-customComponent-totop {
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  position: fixed;
+  z-index: 999;
+  color: @bl-customComponent-totop-themeTextColor;
+  background-color: @bl-customComponent-totop-background-color;
+  border-radius: 50%;
```
Comment (v-excelsior) [inline components/bl-totop-component/styles/index.less:18]: variable for:

border-radius
cursor
z-index

File: components/bl-totop-component/src/index.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  style.backgroundColor = backgroundColor
+  style.padding = padding
+  style.bottom = indent
+
+  if (position === 'bottom-right') {
+    style.right = indent
+  } else {
+    style.left = indent
+  }
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:14]:
```suggestion
const ComponentPosition = {
  BOTTOM_RIGHT: 'bottom-right'
  ...
}

...

const styles = {
  backgroundColor,
  padding,
  bottom: indent,
  right: position === ComponentPosition.BOTTOM_RIGHT ? indent : undefined
  lefth: position !== ComponentPosition.BOTTOM_RIGHT ? indent : undefined
}

```

File: components/bl-totop-component/src/index.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  const iconStyles = {
+    width: size,
+    height: size,
+    fill: color,
+  }
+
+  const handleScroll = () => {
+    if (element && element.el) {
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:23]: if (element?.el) {

File: components/bl-totop-component/src/index.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        top: element.el.offsetTop + offset,
+        behavior: 'smooth',
+      })
+    } else {
+      window.scrollTo({
+        top: offset,
+        behavior: 'smooth',
+      })
+    }
+  }
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:34]:
```suggestion
  const handleScroll = () => window.scrollTo({ top:  element?.el ? offset : offset, behavior: 'smooth' })
  
```

File: components/bl-totop-component/component.json [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      "handlerId": "elementLogic",
+      "handlerLabel": "Element To Scroll To Logic",
+      "handlerDescription": "This is a handler for the logic to determine the element to which the scroll will be."
+    },
+    {
+      "type": "number",
+      "name": "offset",
+      "label": "Offset",
+      "showInSettings": true,
+      "defaultValue": "0",
```
Comment (v-excelsior) [inline components/bl-totop-component/component.json:25]: no need in default value.

Default value should comes from styles(variables)

File: components/bl-totop-component/component.json [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      "handlerId": "sizeLogic",
+      "defaultValue": "24px",
+      "settingTooltip": "The size(width/height) of the icon inside the up button"
+    },
+    {
+      "type": "text",
+      "name": "padding",
+      "label": "Padding",
+      "showInSettings": true,
+      "defaultValue": "10px",
```
Comment (v-excelsior) [inline components/bl-totop-component/component.json:83]: move default value to styles

File: components/bl-totop-component/styles/index.less [hunk 8]
```diff
@@ -0,0 +1,29 @@
+@bl-customComponent-totop-theme: @themePrimary;
+@bl-customComponent-totop-themeTextColor: @appTextColor;
+
+@bl-customComponent-totop-background-color: if(@isLightTheme, rgba(0, 0, 0, 0.04), rgba(255, 255, 255, 0.04));
+@bl-customComponent-totop-hover-background-color: fade(@bl-customComponent-totop-background-color, 20%);
+
+@bl-customComponent-totop-size: 20px;
```
Comment (v-excelsior) [inline components/bl-totop-component/styles/index.less:7]: where we use it?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-totop-component/README.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+**Other**
+````
+@bl-customComponent-totop-cursor: pointer;
+````
+
+## Examples
+
+Below is a Codeless Example highlighting how to use the Totop component:
+
+![totop data example](example-images/totop-data-example.png)
```
Comment (v-excelsior) [inline components/bl-totop-component/README.md:53]: Could we wrap it for centering as thumbnail?

```
<p align="center">
  <img src="./thumbnail.png" alt="main thumbnail" width="780"/>
</p>
```

File: components/bl-totop-component/src/index.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    classList,
+    offset,
+    position,
+    backgroundColor,
+    color,
+    size,
+    iconSize,
+    indent = component.indent ? component.indent : "20px",
+    element,
+  } = component;
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:20]:
```suggestion
  const { 
   classList,    offset,    position,    backgroundColor,    color,    size,    iconSize,    indent =  "20px", element 
   } = component;
```

File: components/bl-totop-component/src/index.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  const styles = {
+    backgroundColor,
+    width: size,
+    height: size,
+    top: position.includes("top") ? indent : position.includes("center-") ? Transform.MAIN : undefined,
+    bottom: position.includes("bottom") ? indent : position.includes("center-") ? Transform.MAIN : undefined,
+    right: position.includes("right") ? indent : position.includes("-center") ? Transform.MAIN : undefined,
+    left: position.includes("left") ? indent : position.includes("-center") ? Transform.MAIN : undefined,
+    transform: position.includes("-center") ? Transform.X : position.includes("center-") ? Transform.Y : undefined,
+  };
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:31]: How about that approach? 
https://getbootstrap.com/docs/5.0/utilities/position/

File: components/bl-totop-component/src/index.js [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    height: iconSize,
+    fill: color,
+  };
+
+  const handleScroll = () => {
+    window.scrollTo({ top:  element?.el ? element.el.offsetTop + offset : offset, behavior: 'smooth' });
+  };
+
+  return (
+    <div className={cn("bl-customComponent-totop", classList)} style={styles} onClick={handleScroll}>
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:44]: Create root ref for component.el property

File: components/bl-totop-component/src/index.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  };
+
+  const iconStyles = {
+    width: iconSize,
+    height: iconSize,
+    fill: color,
+  };
+
+  const handleScroll = () => {
+    window.scrollTo({ top:  element?.el ? element.el.offsetTop + offset : offset, behavior: 'smooth' });
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:40]:
```suggestion
    window.scrollTo({ top:  (element?.el?.offsetTop || 0) + offset, behavior: 'smooth' });
```

File: components/bl-totop-component/src/index.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+  };
+
+  return (
+    <div className={cn("bl-customComponent-totop", classList)} style={styles} onClick={handleScroll}>
+      <TotopIcon iconStyles={iconStyles} />
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:45]: <TotopIcon style={iconStyles} />

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-totop-component/src/index.js (hunk 11 above)
Comment (Fialkaksu) [reply components/bl-totop-component/src/index.js:31]: I think these approaches are roughly equivalent. Applying classes will also require checking the conditions for their application, right there everything you need is immediately collected in styles.

File: components/bl-totop-component/src/index.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  const styles = {
+    backgroundColor,
+    width: size,
+    height: size,
+    top: position.includes("top") ? indent : position.includes("center-") ? Transform.MAIN : undefined,
+    bottom: position.includes("bottom") ? indent : position.includes("center-") ? Transform.MAIN : undefined,
+    right: position.includes("right") ? indent : position.includes("-center") ? Transform.MAIN : undefined,
+    left: position.includes("left") ? indent : position.includes("-center") ? Transform.MAIN : undefined,
+    transform: position.includes("-center") ? Transform.X : position.includes("center-") ? Transform.Y : undefined,
+  };
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:23]:
```
const AxisClassesMap = {
  X: { left: 'start-0', right: 'end-0', center: 'start-50 translate-middle-x' },
  Y: { top: 'top-0', center: 'top-50', bottom: 'bottom-0  translate-middle-y' },
}

const axisPositions = position.split('-')

const classes = useMemo(() => {
  return [...classList, AxisClassesMap.X[axisPositions[0]], AxisClassesMap.Y[axisPositions[0]]]
}, [])
```

Up to 10 cycles over the same array VS 0.

File: components/bl-totop-component/src/index.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+  const iconStyles = {
+    width: iconSize,
+    height: iconSize,
+    fill: color,
+  };
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:29]: useMemo

File: components/bl-totop-component/src/index.js [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    window.scrollTo({ top: (element?.el?.offsetTop || 0) + offset, behavior: "smooth" });
+  };
+
+  const totopRef = useRef();
+
+  component.el = totopRef.current;
+
+  return (
+    <div ref={totopRef} className={cn("bl-customComponent-totop", classList)} style={styles} onClick={handleScroll}>
+      <TotopIcon iconStyles={iconStyles} />
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:41]: TotopIcon already have icon in name, so no need to name props as "iconStyles", use just "styles", and pass iconStyles for it

File: components/bl-totop-component/src/index.js [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    fill: color,
+  };
+
+  const handleScroll = () => {
+    window.scrollTo({ top: (element?.el?.offsetTop || 0) + offset, behavior: "smooth" });
+  };
+
+  const totopRef = useRef();
+
+  component.el = totopRef.current;
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:37]: https://github.com/Backendless/ui-builder-library/pull/195#discussion_r1084203591

>I'm not sure this is the correct way
let's imaging this was rendered only once and the animationOnScrollRef.current before first render will be null and as result it won't work
I assume we need to wrap it with useEffect

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-totop-component/styles/index.less [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+
+
+
+
+
+
+
+
```
Comment (v-excelsior) [inline components/bl-totop-component/styles/index.less:86]: Empty lines

File: components/bl-totop-component/styles/index.less [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+  top: 50%;
+}
+
+.translate-middle-x {
+  transform: translateX(-50%);
+}
+
+.translate-middle-y {
+  transform: translateY(-50%);
+}
```
Comment (v-excelsior) [inline components/bl-totop-component/styles/index.less:77]: That should be under root class

File: components/bl-totop-component/src/index.js [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+
+  const styles = useMemo(() => {
+    return {
+      backgroundColor,
+      width: size,
+      height: size,
+      "--totop-indent-x": indentX,
+      "--totop-indent-y": indentY,
+    };
+  }, []);
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:27]: Read about useMemo please, there should be dependencies.

File: components/bl-totop-component/src/index.js [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+};
+
+export default function TotopComponent({ component }) {
+  const { classList, display, offset, position, backgroundColor, color, size, iconSize, indentX, indentY, element } = component;
+
+  const axisPositions = position.split("-");
+
+  const classes = useMemo(() => {
+    return [...classList, AxisClassesMap.X[axisPositions[1]], AxisClassesMap.Y[axisPositions[0]]];
+  }, []);
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:17]: Read about useMemo please, there should be dependencies.

File: components/bl-totop-component/src/index.js [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+  const handleScroll = () => {
+    window.scrollTo({ top: (element?.el?.offsetTop || 0) + offset, behavior: "smooth" });
+  };
+
+  const totopRef = useRef();
+
+  useEffect(() => {
+    component.el = totopRef.current;
+  }, []);
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:45]: Read about useEffect please, there should be dependencies.

File: components/bl-totop-component/src/index.js [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+
+const { cn } = BackendlessUI.CSSUtils;
+
+const AxisClassesMap = {
+  X: { left: "start-0", right: "end-0", center: "start-50 translate-middle-x" },
+  Y: { top: "top-0", bottom: "bottom-0", center: "top-50  translate-middle-y" },
+};
+
+export default function TotopComponent({ component }) {
+  const { classList, display, offset, position, backgroundColor, color, size, iconSize, indentX, indentY, element } = component;
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:11]: offset, position, backgroundColor, color, size, iconSize, indentX, indentY - we must add logic handlers for that(where it possible)

File: components/bl-totop-component/component.json [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+      "name": "indentY",
+      "label": "Indent Y",
+      "settingTooltip": "Indent of the up button from the edge of the page vertically",
+      "showInSettings": true,
+      "hasLogicHandler": false,
+      "handlerId": "indentYLogic"
+    }
+  ],
+  "eventHandlers": [],
+  "actions": [],
```
Comment (v-excelsior) [inline components/bl-totop-component/component.json:132]: How about actions 'Hide Scroll Button' and 'Show Scroll Button'

It's not really needed, but - it's training for you to work with actions, and they will not interfere

File: components/bl-totop-component/src/index.js [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+
+  component.show = () => {
+    setIsTotopVisible(true);
+  };
+
+  const axisPositions = position.split("-");
+
+  const classes = useMemo(() => {
+    return [...classList, AxisClassesMap.X[axisPositions[1]], AxisClassesMap.Y[axisPositions[0]]];
+  }, [axisPositions]);
```
Comment (v-excelsior) [inline components/bl-totop-component/src/index.js:29]: classList at deps

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Pull request merged**